### PR TITLE
New Resource: `azurerm_management_lock`

### LINKED ISSUE #23: Locks in AzureRM
_This issue was originally opened by @AMMullan as hashicorp/terraform#9768. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi,

I can't see any option for setting Locks on resources, i.e. we want to have a Resource Group in Azure for networking components and have it Locked so only Owners can manage it but Terraform doesn't seem to have this feature?

REST API documentation is here: https://azure.microsoft.com/en-gb/documentation/articles/resource-group-lock-resources/

### Terraform Version
Terraform v0.7.7

### Affected Resource(s)
- azurerm_resource_group
- any other resource that has the Locks options

### Expected Behavior
Create locks for whichever resources need them.

### PR DESCRIPTION
**Pull request by tombuildsstuff**
Note: As the Subscription specific Locks will break other tests; these tests need to be run individually. As such I've introduced the `TF_ACC_SUBSCRIPTION_PARALLEL_LOCK`  environment variable for this purpose - so these tests are run only as needed. Any test failures will be caught by the Resource Group and Resource level tests - so we should be safe here.

Tests pass:

```
$ TF_ACC_SUBSCRIPTION_PARALLEL_LOCK=1 acctests azurerm TestAccAzureRMManagementLock_
=== RUN   TestAccAzureRMManagementLock_importResourceGroupReadOnlyBasic
--- PASS: TestAccAzureRMManagementLock_importResourceGroupReadOnlyBasic (61.52s)
=== RUN   TestAccAzureRMManagementLock_importResourceGroupReadOnlyComplete
--- PASS: TestAccAzureRMManagementLock_importResourceGroupReadOnlyComplete (58.75s)
=== RUN   TestAccAzureRMManagementLock_importResourceGroupCanNotDeleteBasic
--- PASS: TestAccAzureRMManagementLock_importResourceGroupCanNotDeleteBasic (53.38s)
=== RUN   TestAccAzureRMManagementLock_importResourceGroupCanNotDeleteComplete
--- PASS: TestAccAzureRMManagementLock_importResourceGroupCanNotDeleteComplete (46.87s)
=== RUN   TestAccAzureRMManagementLock_importPublicIPCanNotDeleteBasic
--- PASS: TestAccAzureRMManagementLock_importPublicIPCanNotDeleteBasic (80.46s)
=== RUN   TestAccAzureRMManagementLock_importPublicIPReadOnlyBasic
--- PASS: TestAccAzureRMManagementLock_importPublicIPReadOnlyBasic (68.53s)
=== RUN   TestAccAzureRMManagementLock_resourceGroupReadOnlyBasic
--- PASS: TestAccAzureRMManagementLock_resourceGroupReadOnlyBasic (61.24s)
=== RUN   TestAccAzureRMManagementLock_resourceGroupReadOnlyComplete
--- PASS: TestAccAzureRMManagementLock_resourceGroupReadOnlyComplete (64.10s)
=== RUN   TestAccAzureRMManagementLock_resourceGroupCanNotDeleteBasic
--- PASS: TestAccAzureRMManagementLock_resourceGroupCanNotDeleteBasic (72.49s)
=== RUN   TestAccAzureRMManagementLock_resourceGroupCanNotDeleteComplete
--- PASS: TestAccAzureRMManagementLock_resourceGroupCanNotDeleteComplete (113.71s)
=== RUN   TestAccAzureRMManagementLock_publicIPReadOnlyBasic
--- PASS: TestAccAzureRMManagementLock_publicIPReadOnlyBasic (64.05s)
=== RUN   TestAccAzureRMManagementLock_publicIPCanNotDeleteBasic
--- PASS: TestAccAzureRMManagementLock_publicIPCanNotDeleteBasic (94.53s)
=== RUN   TestAccAzureRMManagementLock_subscriptionReadOnlyBasic
--- PASS: TestAccAzureRMManagementLock_subscriptionReadOnlyBasic (17.98s)
=== RUN   TestAccAzureRMManagementLock_subscriptionCanNotDeleteBasic
--- PASS: TestAccAzureRMManagementLock_subscriptionCanNotDeleteBasic (15.20s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm    872.839s
```

Fixes #23

### CODE REVIEW

File: azurerm/resource_arm_management_lock.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				// TODO: validation
```
Comment (mbfrahry) [inline azurerm/resource_arm_management_lock.go]: Do we want to add validation with this PR?

File: azurerm/resource_arm_management_lock.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	return nil
+}
+
+type AzureManagementLockId struct {
+	Scope string
+	Name  string
+}
+
+func parseAzureRMLockId(id string) (*AzureManagementLockId, error) {
+	segments := strings.Split(id, "/providers/Microsoft.Authorization/locks/")
```
Comment (mbfrahry) [inline azurerm/resource_arm_management_lock.go]: Thoughts on confirming we have two parts?

**Review by mbfrahry — CHANGES_REQUESTED** [review]
You mentioned that `TF_ACC_SUBSCRIPTION_PARALLEL_LOCK` needs to be set for these tests to run so it doesn't break other tests but only a couple of the tests check for this. Do only those two tests need this or do all them?

**Comment by tombuildsstuff** [comment]
> You mentioned that TF_ACC_SUBSCRIPTION_PARALLEL_LOCK needs to be set for these tests to run so it doesn't break other tests but only a couple of the tests check for this. Do only those two tests need this or do all them?

Only these two tests do (since they lock the Subscription - which means no other resources within the Subscription can be edited) - the other tests are limited to a single Resource Group / Resource - so they can be run as needed.

**Comment by mbfrahry** [comment]
Good to know!

**Review by mbfrahry — APPROVED** [review]
LGTM after discussing

File: azurerm/resource_arm_management_lock.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_management_lock.go]: fixed

File: azurerm/resource_arm_management_lock.go (hunk 1 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_management_lock.go]: done

**Pull request merged**